sql/atxs/atxs.go

```diff
@@ -0,0 +1,213 @@
+		stmt.BindBytes(2, nodeID.ToBytes())
+	}
+	dec := func(stmt *sql.Statement) bool {
+		stmt.ColumnBytes(0, id[:])
+		return true
+	}
+
+	if rows, err := db.Exec(`
+		select id from atxs 
+		where epoch = ?1 and smesher = ?2 
+		order by timestamp desc 
+		limit 1;`, enc, dec); err != nil {
+		return *types.EmptyATXID, fmt.Errorf("exec id %v: %w", id, err)
+	} else if rows == 0 {
+		return *types.EmptyATXID, fmt.Errorf("exec id %s: %w", id, sql.ErrNotFound)
+	}
+
+	return id, err
+}
+
+// GetIDsByEpoch gets ATX IDs for a given epoch.
+func GetIDsByEpoch(db sql.Executor, epoch types.EpochID) (ids []types.ATXID, err error) {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindInt64(1, int64(epoch))
+	}
+	dec := func(stmt *sql.Statement) bool {
+		var id types.ATXID
+		stmt.ColumnBytes(0, id[:])
+		ids = append(ids, id)
+		return true
+	}
+
+	if rows, err := db.Exec(`
+		select id from atxs 
+		where epoch = ?1 
+		order by id asc;`, enc, dec); err != nil {
+		return nil, fmt.Errorf("exec epoch %v: %w", epoch, err)
+	} else if rows == 0 {
+		return []types.ATXID{}, nil
+	}
+
+	return ids, nil
+}
+
+// GetTop gets a top ATX (positioning ATX candidate) for a given ATX ID.
+// The top ATX is the one with the highest layer ID.
+func GetTop(db sql.Executor) (id types.ATXID, err error) {
+	dec := func(stmt *sql.Statement) bool {
+		stmt.ColumnBytes(0, id[:])
+		return true
+	}
+
+	if rows, err := db.Exec("select id from atxs order by layer desc, timestamp asc limit 1;", nil, dec); err != nil {
+		return *types.EmptyATXID, fmt.Errorf("exec: %w", err)
+	} else if rows == 0 {
+		return *types.EmptyATXID, fmt.Errorf("exec: %w", sql.ErrNotFound)
+	}
+
+	return id, err
+}
+
+// Add adds an ATX for a given ATX ID.
+func Add(db sql.Executor, id types.ATXID, atx *types.ActivationTx, timestamp time.Time) error {
+	buf, err := codec.Encode(atx)
+	if err != nil {
+		return fmt.Errorf("encode: %w", err)
+	}
+
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, id.Bytes())
+		stmt.BindInt64(2, int64(atx.PubLayerID.Uint32()))
+		stmt.BindInt64(3, int64(atx.PubLayerID.GetEpoch()))
+		stmt.BindBytes(4, atx.NodeID.ToBytes())
+		stmt.BindBytes(5, buf)
+		stmt.BindInt64(6, timestamp.UnixNano())
+	}
+
+	_, err = db.Exec(`
+		insert into atxs (id, layer, epoch, smesher, atx, timestamp) 
+		values (?1, ?2, ?3, ?4, ?5, ?6);`, enc, nil)
+	if err != nil {
+		return fmt.Errorf("insert ATX ID %v: %w", id, err)
+	}
+
+	return nil
+}
+
+// DeleteATXsByNodeID deletes ATXs by node ID.
+func DeleteATXsByNodeID(db sql.Executor, nodeID types.NodeID) error {
+	enc := func(stmt *sql.Statement) {
+		stmt.BindBytes(1, nodeID.ToBytes())
+	}
+
+	if _, err := db.Exec("delete from atxs where smesher = ?1;", enc, nil); err != nil {
+		return fmt.Errorf("exec %v: %w", nodeID, err)
+	}
+
+	return nil
+}
+
+// FetchAdapter is an adapter of SQLite implementation to legacy LevelDB interfaces.
+// TODO(nkryuchkov): Remove when transition to SQLite is finished.
+type FetchAdapter struct {
```

Comment:
lets move it to the other module, maybe activations, so that the approach is consistent with mesh module